File: services/frontend/src/lib/api/schools/get-schools.ts
import { convertStructTypes } from "$lib/functions/convertStructTypes";
import type { SchoolShortResponseModelType } from "./types/school-short-model-type";

export async function getSchools(page: number, filter: string): Promise<SchoolShortResponseModelType | undefined> {
    try {

        const queryParams = new URLSearchParams();
        queryParams.append('page', page.toString());
        if (filter) {
            queryParams.append('filterResult', filter);
        }

        const response = await fetch(`/api/v1/schools?${queryParams.toString()}`);
        if (!response.ok) {
            throw new Error('Failed to fetch school data');
        }

        const rawData = await response.json()
        const data = convertStructTypes(rawData)

        return data;
    } catch (error) {

        console.error('Failed to fetch school data', error);
        return undefined
    }
}

File: services/backend/src/resources/schools/http/get_schools.go
package http

import (
	"net/http"

	customerrorconst "github.com/dcssoftware/bafoeg-manager/src/helper/debug/customerrors/custom-error-const"
	httpParams "github.com/dcssoftware/bafoeg-manager/src/helper/http-params"
	"github.com/dcssoftware/bafoeg-manager/src/resources/schools/http/models"
	"github.com/gofiber/fiber/v3"
)

//	@Summary		V1 Get schools
//	@Description	Retrieves V1 Schools
//	@Tags			schools
//	@Produce		json
//	@Param			page	query		int		true	"Page number for pagination"
//
// Success 		200 {object} 	models.SchoolShortResponseModel "School Model"
//
//	@Failure		404		{string}	string	"The requested data could not be found."
//	@Failure		500		{string}	string	"An error occurred while processing your request. Please try again later. "
//
//	@Router			/api/v1/schools/ [get]
func (h *SchoolHandler) GetSchools(c fiber.Ctx) error {

	queries := c.Queries()
	filterResultString := queries["filterResult"]
	pageNumber, pageNumberErr := httpParams.GetParamsPageUint(c.Queries())
	if pageNumberErr != nil {
		return c.Status(http.StatusBadRequest).SendString(customerrorconst.BAD_REQUEST_ERROR_MESSAGE)
	}

	schools, schoolsErr := h.service.GetSchools(nil, pageNumber, filterResultString)
	if schoolsErr != nil {
		status, message := schoolsErr.HTTPError()
		return c.Status(status).SendString(message)
	}

	schoolsCount, schoolsCountErr := h.service.GetSchoolsCount(nil, filterResultString)
	if schoolsCountErr != nil {
		status, message := schoolsCountErr.HTTPError()
		return c.Status(status).SendString(message)
	}

	httpSchools := models.ToHttpSchoolShortModels(schools, schoolsCount)
	return c.Status(http.StatusOK).JSON(httpSchools)
}
